discojs-web/src/memory/memory.ts

```diff
@@ -10,62 +10,72 @@ import { Map } from 'immutable'
 import * as tf from '@tensorflow/tfjs'
 
 import type { Path, Model, ModelInfo, ModelSource } from '@epfml/discojs'
-import { Memory, ModelType, models } from '@epfml/discojs'
+import { Memory, StoredModelType, models } from '@epfml/discojs'
 
 export class IndexedDB extends Memory {
-  override pathFor (source: ModelSource): Path {
+  override getModelMemoryPath (source: ModelSource): Path {
     if (typeof source === 'string') {
       return source
     }
-
-    if (source.type === undefined || source.taskID === undefined || source.name === undefined) {
+    if (source.type === undefined || source.taskID === undefined ||
+      source.name === undefined || source.tensorBackend === undefined) {
       throw new TypeError('source incomplete')
     }
-
     const version = source.version ?? 0
-
-    return `indexeddb://${source.type}/${source.taskID}/${source.name}@${version}`
+    return `indexeddb://${source.type}/${source.tensorBackend}/${source.taskID}/${source.name}@${version}`
   }
 
-  override infoFor (source: ModelSource): ModelInfo {
+  override getModelInfo (source: ModelSource): ModelInfo {
     if (typeof source !== 'string') {
       return source
     }
-    const [stringType, taskID, fullName] = source.split('/').splice(2)
+    const [stringType, tensorBackend, taskID, fullName] = source.split('/').splice(2)
 
-    const type = stringType === 'working' ? ModelType.WORKING : ModelType.SAVED
+    const type = stringType === 'working' ? StoredModelType.WORKING : StoredModelType.SAVED
 
     const [name, versionSuffix] = fullName.split('@')
     const version = versionSuffix === undefined ? 0 : Number(versionSuffix)
-    return { type, taskID, name, version }
+    if (tensorBackend != 'tfjs' && tensorBackend != 'gpt') {
+      throw Error("Unknown tensor backend")
+    }
+    return { type, taskID, name, version, tensorBackend }
   }
 
   async getModelMetadata (source: ModelSource): Promise<tf.io.ModelArtifactsInfo | undefined> {
     const models = await tf.io.listModels()
-    return models[this.pathFor(source)]
+    return models[this.getModelMemoryPath(source)]
   }
 
   async contains (source: ModelSource): Promise<boolean> {
     return await this.getModelMetadata(source) !== undefined
   }
 
-  override async getModel (source: ModelSource): Promise<Model> {
-    return new models.TFJS(await tf.loadLayersModel(this.pathFor(source)))
+  override async getModel(source: ModelSource): Promise<Model> {
+    const layersModel = await tf.loadLayersModel(this.getModelMemoryPath(source))
+    
+    const tensorBackend = this.getModelInfo(source).tensorBackend
+    if (tensorBackend == 'tfjs') {
+      return new models.TFJS(layersModel)
+    } else if (tensorBackend == 'gpt') {
+      return new models.GPT(undefined, layersModel)
+    } else {
+      throw Error("tensor backend unrecognized")
+    }
```

Comment:
there is support for [exaustiveness checking](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking), which allow TS to know that every possible value was checked. a neat feature when changing the type of `tensorBackend` (as it will make compilation fail instead of throwing at runtime), and reads more nicely IMO

```suggestion
    switch (tensorBackend) {
      case 'tfjs':
        return new models.TFJS(layersModel)
      case 'gpt':
        return new models.GPT(undefined, layersModel)
    }
```

Reply:
Good to know thanks!